### PR TITLE
[mono] Use InlineArrayAttribute for ArgumentData and StackAllocatedByRefs

### PR DESCRIPTION
**Pull request by lambdageek**
Mono supports InlineArray

### CODE REVIEW

**Comment by jkotas** [comment]
There is one more src\libraries\System.Private.CoreLib\src\System\ParamsArray.cs

File: src/libraries/System.Private.CoreLib/src/System/Reflection/MethodBase.cs [hunk 1]
```diff
@@ -274,19 +274,11 @@ BindingFlags invokeAttr
 
         internal const int MaxStackAllocArgCount = 4;
 
-#if CORECLR
+
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Reflection/MethodBase.cs:277]:
```suggestion
```

File: src/libraries/System.Private.CoreLib/src/System/Reflection/MethodBase.cs [hunk 2]
```diff
@@ -274,19 +274,11 @@ BindingFlags invokeAttr
...
         [InlineArray(MaxStackAllocArgCount)]
-#endif
         private protected struct ArgumentData<T>
         {
             private T _arg0;
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Reflection/MethodBase.cs:281]:
```suggestion
            private T _arg0;

```

**Review by jkotas — APPROVED** [review]
Thanks!

**Comment by uweigand** [comment]
Hi @lambdageek, this seems to have caused a severe regression on s390x: *all* CI tests are currently failing with this error:
```
error: Could not set up field '_args' due to: Generic type definition failed, due to: Inline array struct size out of bounds, abnormally large.
```
I haven't looked into the root cause yet - it looks like this is failing only on s390x, not ppc64le, so maybe an endian issue?

**Comment by uweigand** [comment]
I see the following suspicious code in has_inline_array_attribute_value_func:
```
                attr->value = *(gpointer*)decoded_attr->typed_args [0]->value.primitive;
```
and later
```
                        klass->inlinearray_value = GPOINTER_TO_INT32 (attr.value);
```

If this is actually a 4-byte value in memory, then the `*(gpointer*)` is broken on big-endian.

**Comment by uweigand** [comment]
And indeed the following patch fixes the problem for me:
```
diff --git a/src/mono/mono/metadata/class-init.c b/src/mono/mono/metadata/class-init.c
index c41ea7b891d..ba5954e7963 100644
--- a/src/mono/mono/metadata/class-init.c
+++ b/src/mono/mono/metadata/class-init.c
@@ -816,7 +816,7 @@ has_inline_array_attribute_value_func (MonoImage *image, uint32_t method_token,
                MonoDecodeCustomAttr *decoded_attr = mono_reflection_create_custom_attr_data_args_noalloc (image, ctor, (guchar*)data, data_size, &error);
                mono_error_assert_ok (&error);
                g_assert (decoded_attr->named_args_num == 0 && decoded_attr->typed_args_num == 1);
-               attr->value = *(gpointer*)decoded_attr->typed_args [0]->value.primitive;
+               attr->value = (gpointer)*(guint32*)decoded_attr->typed_args [0]->value.primitive;
                g_free (decoded_attr);
        } else {
                g_warning ("Can't find custom attr constructor image: %s mtoken: 0x%08x due to: %s", image->name, method_token, mono_error_get_message (&error));
```
Does this look reasonable to you?

**Comment by lambdageek** [comment]
Maybe we should use one of our READ32 macros (sorry, I'm on a phone, can't look up the exact name)

**Comment by uweigand** [comment]
Looks like READ32 is only available/used inside the interpreter (src/mono/mini/interp), not in the rest of Mono.

**Comment by uweigand** [comment]
I've now opened a PR including this fix: https://github.com/dotnet/runtime/pull/88417

**Pull request merged**